membershipextras.php

```diff
@@ -166,22 +166,48 @@ function membershipextras_civicrm_pre($op, $objectName, $id, &$params) {
     $preEditMembershipHook->preventExtendingOfflinePendingRecurringMembership();
   }
 
-  $action = CRM_Utils_Request::retrieve('action', 'String');
-  $context = CRM_Utils_Request::retrieve('context', 'String');
+  $isPaymentPlanPayment = _membershipextras_isPaymentPlanPayment();
 
-  $membershipContributionCreation = $objectName === 'Contribution' && $op === 'create'
-    && $context === 'membership';
-  $newMembershipContribution = $membershipContributionCreation && ($action & CRM_Core_Action::ADD);
-  $renewMembershipContribution = $membershipContributionCreation && ($action & CRM_Core_Action::RENEW);
+  $membershipContributionCreation = ($objectName === 'Contribution' && $op === 'create' && !empty($params['membership_id']));
+  static $isUpfrontContribution = TRUE;
+  if ($membershipContributionCreation && $isPaymentPlanPayment && $isUpfrontContribution) {
```

Comment:
I don't think I understand what the purpose of $isUpFrontContribution is. We set it right before we evaluate it, and if the hook is run for a membership creation for a payment plan, then it is then set to false. But it is never checked again. Am I missing something?

Reply:
Sorry, I used the wrong variable names here by mistake, this should be $isFirstPaymentPlanContribution  instead of $isUpFrontContribution (I will fix it in my next commit).

But to explain it, as you know civicrm will be creates the first contribtuion by default but the amount of the contribution and line items will not be correct for it when we pay using payment plan option.
When this hook get called for the first time and the payment is done using payment plan option, then we know for sure that the its get called before creating the first payment plan contribtuion, so we create the recurring contribution and update the first contribution amounts using  :

```php
$paymentPlanProcessor = new CRM_MembershipExtras_Hook_Pre_MembershipPaymentPlanProcessor($params);
 $paymentPlanProcessor->createPaymentPlan();
```

then I set $isFirstPaymentPlanContribution to false after doing that, because when the rest of the payment plan contributions get created, we don't want to alter their amounts since we create and calculate their amounts correctly in our extension before creating them so altering their amounts is not needed. Hence that this variable is static, so each time the hook is called the previous value will be preserved resulting on only calling this line : 

```php
$paymentPlanProcessor = new CRM_MembershipExtras_Hook_Pre_MembershipPaymentPlanProcessor($params);
 $paymentPlanProcessor->createPaymentPlan();
```

once and only for the first payment plan contribution.